### ISSUE TITLE
no-use-before-define triggers for variables defined below a function even if variable will be defined when a function using that variable is called.

### ISSUE THREAD

**Issue by AriaFallah** [eslint/eslint#6415]
**What version of ESLint are you using?**
2.12.0

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**
https://github.com/AriaFallah/eslint-config-aria/blob/master/index.js

Specific line:

```json
"no-use-before-define": [2, "nofunc"]
```

**What did you do? Please include the actual source code causing the issue.**

```js
function x() {
  return myFunc()
}

const myFunc = function() {
  return 5
}

console.log(x())
```

**What did you expect to happen?**
No errors. This doesn't result in `myFunc` is undefined in console because the function is not called before myFunc is defined.

**What actually happened? Please include the actual, raw output from ESLint.**
```
error  'myFunc' was used before it was defined  no-use-before-define
```

**Comment by platinumazure** [comment]
According to the [rule options documentation](http://eslint.org/docs/rules/no-use-before-define#options), the correct syntax for the rule options should look like this:

```json
{
    "rules": {
        "no-use-before-define": [2, { "functions": false }]
    }
}
```

Try that and see if it works for you.

**Comment by AriaFallah** [comment]
@platinumazure 

From the docs that you have linked yourself

> This rule accepts "nofunc" string as a option. "nofunc" is the same as 
> `{ "functions": false, "classes": true }`.

**Comment by vsemozhetbyt** [comment]
@AriaFallah

```js
const myFunc = function() {
  return 5
}
```
This is not function declaration, this is assignment of function expression.

**Comment by AriaFallah** [comment]
@vsemozhetbyt 

Which is probably why eslint is throwing the error. Perhaps I should change the title so it's not function centric, but rather addresses variables declared below a function that uses that variable.

**Comment by vsemozhetbyt** [comment]
@AriaFallah But such a behavior is the rule destination. Еру rule exceptions concern only function and class declarations, not assignment. You really try to use-before-define the variable `myFunc`.

**Comment by AriaFallah** [comment]
@vsemozhetbyt 

I guess it's an edge case and perhaps not a big deal, but it is defined if you call the function `x` after `myFunc` has been defined. [Even though it is specified as `const` it is still hoisted to the top of the scope](http://stackoverflow.com/questions/31219420/are-variables-declared-with-let-or-const-not-hoisted-in-es6).

The only reason this is bothering me is because there's no nice way to decorate function declarations without breaking either [no-func-assign](http://eslint.org/docs/rules/no-func-assign) or `no-use-before-define`.

```js
function x() {
  y()
}

function y() {
}
y = decorate(y) // breaks no-func-assign
```

or

```
function x() {
  y() // breaks no-use-before-define
}

const y = decorate(function x() {})
```

**Comment by vsemozhetbyt** [comment]
@AriaFallah See this difference in the answer:

> The difference between var/function/function* declarations and let/const/class declara­tions is the initialisation.
The former are initialised with undefined or the (generator) function right when the binding is created at the top of the scope. The lexically declared variables however stay uninitialised. This means that a ReferenceError exception is thrown when you try to access it. It will only get initialised when the let/const/class statement is evaluated, everything above that is called the temporal dead zone.

**Comment by vsemozhetbyt** [comment]
I see.

**Comment by AriaFallah** [comment]
@vsemozhetbyt 

Right, but that's only if they're in the same scope. If you run the example code I provided in the babel REPL, you'll get:

![screen shot 2016-06-14 at 4 56 48 pm](https://cloud.githubusercontent.com/assets/9424976/16059469/03b913a6-3251-11e6-9706-2ff9320baec6.png)

**Comment by ilyavolodin** [comment]
@AriaFallah The point of this rule is to not use variables before they are declared. The example that you provided does exactly what this rule is supposed to catch. Exclusion for function declarations was made because function declarations are hoisted, and it's safe to use them. So are variables, however, again, point of this rule is not to verify validity of your JavaScript, but to enforce convention. It looks like your personal convention goes against convention that this rule is trying to enforce. It seems that you should turn this rule off in your global configuration.

P.S. Exception for function declarations was actually done for recursion, as there was no way to write recurring functions without tripping this rule up.

**Comment by platinumazure** [comment]
Closing as question seems to have been answered. Please reopen if there is more to discuss.
